### PR TITLE
Work around Samsung keyboard issue

### PR DESCRIPTION
**Pull request by mklim**
Samsung's Korean keyboard has a bug where it always attempts to combine
characters based on its internal state, ignoring if and when the cursor
is moved programmatically. EG typing "ㄴㅇ" and then moving the cursor
back to the front of the text and typing "ㄴ" again would result in
"ㄴㅇㄴ", not "ㄴㄴㅇ".

Fully restarting the IMM works around this because it flushes the
keyboard's internal state and stops it from trying to incorrectly
combine characters. However this also has some negative performance
implications, so we only apply the workaround on Samsung devices set
to use Korean input.

This also effectively disables the feature on Samsung keyboards that
allowed users to re-open a composing region for previously typed
characters. See https://github.com/flutter/flutter/issues/29341#issuecomment-531283508.

Fixes flutter/flutter#29341.

### CODE REVIEW

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 1]
```diff
@@ -308,6 +313,20 @@ private void setTextInputEditingState(View view, TextInputChannel.TextEditState
...
     }
 
+    // Samsung's Korean keyboard has a bug where it always attempts to combine characters based on
+    // its internal state, ignoring if and when the cursor is moved programmatically.
+    //
+    // Fully restarting the IMM works around this because it flushes the keyboard's internal state
+    // and stops it from trying to incorrectly combine characters. However this also has some
+    // negative performance implications, so we don't want to apply this workaround in every case.
+    @SuppressLint("NewApi") // New API guard is inline, the linter can't see it.
+    private boolean isRestartAlwaysRequired() {
```
Comment (GaryQian) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: Would it be beneficial to cache the results of this? Not sure if and of these checks requires disk IO as you mentioned checking for keyboard may require.

Comment (mklim) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: Yeah, I think it makes sense. I'm currently caching the result as a final member in the constructor.

Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: Just kidding you already do so!

**Comment by GaryQian** [comment]
I'll ask @jayoung-lee for an opinion on if the lack of capability to reopen composing regions is significant or not.

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 2]
```diff
@@ -293,8 +298,8 @@ private void applyStateToSelection(TextInputChannel.TextEditState state) {
         }
     }
 
-    private void setTextInputEditingState(View view, TextInputChannel.TextEditState state) {
-        if (!mRestartInputPending && state.text.equals(mEditable.toString())) {
+    @VisibleForTesting void setTextInputEditingState(View view, TextInputChannel.TextEditState state) {
+        if (!restartAlwaysRequired && !mRestartInputPending && state.text.equals(mEditable.toString())) {
```
[the remark this answers is not in the document]
Comment (mklim) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: Good question. From testing and reading through the code and manually testing it out it looked like just tapping to move, but I'm not totally convinced that I'm seeing the whole situation here. @jason-simmons do you know?

[the remark this answers is not in the document]
Comment (jason-simmons) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: The engine asks the IMM to restart input when the app switches to a new text editing client or when the app sends an update to editor state originating from the Dart side.

In the latter case it's possible that we could diff the current and new state and do something less expensive than a full restart if the diff is simple enough.  But AFAIK this hasn't been noticeable.

**Comment by jayoung-lee** [comment]
I helped @justinmc play around with the new keyboard, and yes, it seemed to work perfectly fine!  

Please let me know if anything new comes up in the future, I will be more than happy to help you test it :)

**Comment by mklim** [comment]
Thank you @jayoung-lee!

**Comment by futterer** [comment]
Might this be related to [flutter/flutter#41315](https://github.com/flutter/flutter/issues/41315) ?

**Comment by markmoskalenko** [comment]
Included in the assembly of the Russian language!
Similar problem

**Pull request merged**